cdc/generation.cc

```diff
@@ -154,137 +158,69 @@ topology_description generate_topology_description(
-
-            // Find the token range into which our stream_id's token landed.
-            auto it = std::lower_bound(tokens.begin(), tokens.end(), token);
-            auto& entry = entries[it != tokens.end() ? std::distance(tokens.begin(), it) : 0];
-
-            auto shard_id = dht::shard_of(entry.streams.size(), entry.sharding_ignore_msb, token);
-            assert(shard_id < entry.streams.size());
-
-            if (!entry.streams[shard_id].is_set()) {
-                --spots_to_fill;
-                entry.streams[shard_id] = stream_id;
-            }
+        dht::sharder sharder(desc.streams.size(), desc.sharding_ignore_msb);
+        for (size_t shard_idx = 0; shard_idx < desc.streams.size(); ++shard_idx) {
+            auto t = get_token(sharder, shard_idx);
+            desc.streams[shard_idx] = create_stream_id(t);
         }
-
-        if (!spots_to_fill) {
-            return stop_iteration::yes;
+    };
+
+    // First we special case vnode (tokens.back(), tokens.front()] because it wraps around
+    setup_vnode(entries[0], tokens.back(), tokens.front(), [start = tokens.back(), end = tokens.front()] (const dht::sharder& sharder, size_t shard_idx) {
+        auto t = dht::token::from_int64(dht::token::to_int64(start) == std::numeric_limits<int64_t>::max()
+                ? std::numeric_limits<int64_t>::min() : dht::token::to_int64(start) + 1);
+        if (sharder.shard_of(t) != shard_idx) {
+            t = sharder.token_for_next_shard(start, shard_idx);
         }
-
-        auto now = std::chrono::system_clock::now();
-        auto passed = std::chrono::duration_cast<std::chrono::seconds>(now - start_time);
-        if (passed > quota) {
-            return stop_iteration::yes;
+        if (!t.is_maximum()) {
+            return t;
         }
-
-        return stop_iteration::no;
-    }).get();
-
-    if (spots_to_fill) {
-        // We were not able to generate stream ids for each (token range, shard) pair.
-
-        // For each range that has a stream, for each shard for this range that doesn't have a stream,
-        // use the stream id of the next shard for this range.
-
-        // For each range that doesn't have any stream,
-        // use streams of the first range to the left which does have a stream.
-
-        cdc_log.warn("Generation of CDC streams failed to create streams for some (vnode, shard) pair."
-                     " This can lead to worse performance.");
-
-        stream_id some_stream;
-        size_t idx = 0;
-        for (; idx < entries.size(); ++idx) {
-            for (auto s: entries[idx].streams) {
-                if (s.is_set()) {
-                    some_stream = s;
-                    break;
-                }
-            }
-            if (some_stream.is_set()) {
-                break;
-            }
+        t = dht::token::from_int64(std::numeric_limits<int64_t>::min());
+        if (sharder.shard_of(t) != shard_idx) {
+            t = sharder.token_for_next_shard(t, shard_idx);
```

Comment:
I have absolutely no idea what this token/sharder wizardry does.

If I understand correctly, the problem statement is:
given token `start`, find the smallest token "after" `start` which falls into shard `shard_idx`.

Can you:
1. take this piece of code and name it (inside a pure static function), e.g. with the following signature
```
static dht::token first_token_for_shard(dht::token start, shard_id shard, sharder&)
```
2. write a comment outside the function explaining what the function does and how it's different from `token_for_next_shard`
3. write a comment inside the function explaining how the function does it (why do we need all this magic)?

Reply:
I will extract the logic to free functions and add some comments.

Reply:
This is already implemented by ring_position_range_vector_sharder. The duplication isn't helping anything.